lib/types.d.ts

```diff
@@ -13,28 +33,119 @@ type TypeOptions = MergeBy<
+ */
 type Key = TypeOptions['keys'];
+
+/**
+ * Represents a string union of all Onyx collection keys.
+ */
 type CollectionKeyBase = TypeOptions['collectionKeys'];
+
+/**
+ * Represents a literal string union of all Onyx collection keys.
+ * It allows appending a string after each collection key e.g. `report_some-id`.
+ */
 type CollectionKey = `${CollectionKeyBase}${string}`;
 
+/**
+ * Represents a string union of all Onyx normal and collection keys.
+ */
 type OnyxKey = Key | CollectionKey;
+
+/**
+ * Represents a Record where each key is an Onyx key and each value is its corresponding Onyx value type.
+ */
 type KeyValueMapping = TypeOptions['values'];
 
-type GetOnyxValue<K extends Key | `${CollectionKeyBase}${string}`> = K extends Key
-    ? KeyValueMapping[K]
-    : K extends `${CollectionKeyBase}`
-    ? Record<string, KeyValueMapping[K] | null>
-    : KeyValueMapping[K];
+/**
+ * Represents a selector function type which operates based on the provided `TKey` and `ReturnType`.
+ *
+ * A `Selector` is a function that accepts a value and returns a processed value.
+ * This type accepts two type parameters: `TKey` and `TReturnType`.
+ *
+ * The type `TKey` extends `OnyxKey` and it is the key used to access a value in `KeyValueMapping`.
+ * `TReturnType` is the type of the returned value from the selector function.
+ *
+ * If `KeyValueMapping[TKey]` is equal to 'unknown', the `Selector` type represents a function
+ * that takes any value and returns an `unknown` value.
+ *
+ * If `KeyValueMapping[TKey]` is not 'unknown', the `Selector` type represents a function that
+ * takes either the value of type `OnyxEntry<KeyValueMapping[TKey]>`, and returns a value of `TReturnType`.
+ */
+type Selector<TKey extends OnyxKey, TReturnType> = IsEqual<KeyValueMapping[TKey], unknown> extends true
+    ? (value: unknown) => unknown
+    : (value: OnyxEntry<KeyValueMapping[TKey]>) => TReturnType;
 
-// type Selector<TKey extends OnyxKey, TReturnType> = KeyValueMapping[TKey] extends object | string | number | boolean
-//     ? (value: KeyValueMapping[TKey] | null) => TReturnType
-//     : KeyValueMapping[TKey] extends unknown
-//     ? (value: unknown) => unknown
-//     : never;
+/**
+ * Represents a single Onyx entry, that can be either `TOnyxValue` or `null` if it doesn't exist.
+ */
+type OnyxEntry<TOnyxValue> = TOnyxValue | null;
 
-// type Selector<TKey extends OnyxKey, TReturnType> = IsEqual<KeyValueMapping[TKey], unknown> extends true ? (value: unknown) => unknown : (value: KeyValueMapping[TKey] | null) => TReturnType;
-type Selector<TKey extends OnyxKey, TReturnType> = IsEqual<GetOnyxValue<TKey>, unknown> extends true ? (value: unknown) => unknown : (value: GetOnyxValue<TKey> | null) => TReturnType;
+/**
+ * Represents an Onyx collection of entries, that can be either a record of `TOnyxValue`s or `null` if it is empty or doesn't exist.
+ */
+type OnyxCollectionEntries<TOnyxValue> = OnyxEntry<Record<string, TOnyxValue | null>>;
```

Comment:
NAB I prefer calling it `OnyxCollection`.

Reply:
@neil-marcellini Makes sense, updated!